// @tiptap/extension-hard-break/dist/tiptap-extension-hard-break.umd.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tiptap/core')) :
  typeof define === 'function' && define.amd ? define(['exports', '@tiptap/core'], factory) :
  (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global['@tiptap/extension-hard-break'] = {}, global.core));
}(this, (function (exports, core) { 'use strict';

  const HardBreak = core.Node.create({
      name: 'hardBreak',
      defaultOptions: {
          HTMLAttributes: {},
      },
      inline: true,
      group: 'inline',
      selectable: false,
      parseHTML() {
          return [
              { tag: 'br' },
          ];
      },
      renderHTML({ HTMLAttributes }) {
          return ['br', core.mergeAttributes(this.options.HTMLAttributes, HTMLAttributes)];
      },
      addCommands() {
          return {
              setHardBreak: () => ({ commands }) => {
                  return commands.first([
                      () => commands.exitCode(),
                      () => commands.insertContent({ type: this.name }),
                  ]);
              },
          };
      },
      addKeyboardShortcuts() {
          return {
              'Mod-Enter': () => this.editor.commands.setHardBreak(),
              'Shift-Enter': () => this.editor.commands.setHardBreak(),
          };
      },
  });

  exports.HardBreak = HardBreak;
  exports.default = HardBreak;

  Object.defineProperty(exports, '__esModule', { value: true });

})));
